carsons/carsons.py

```diff
@@ -183,3 +181,83 @@ def calculate_distance(positionᵢ, positionⱼ):
     def get_h(self, i):
         _, yᵢ = self.phase_positions[i]
         return yᵢ
+
+    @property
+    def impedance(self):
+        z_primitive = self.build_z_primitive()
+        z_abc = perform_kron_reduction(z_primitive)
+        return z_abc
+
+
+class ConcentricNeutralCarsonsEquations(CarsonsEquations):
+    def __init__(self, model, *args, **kwargs):
+        super().__init__(model)
+        self.neutral_strand_gmr = model.neutral_strand_gmr
+        self.neutral_strand_count = defaultdict(
+            lambda: None, model.neutral_strand_count)
+        self.neutral_strand_resistance = model.neutral_strand_resistance
+        self.radius = defaultdict(lambda: None, {
+            phase: (diameter_over_neutral -
+                    model.neutral_strand_diameter[phase]) / 2
+            for phase, diameter_over_neutral
+            in model.diameter_over_neutral.items()
+        })
+        self.phase_positions.update({
+            f"N{phase}": self.phase_positions[phase]
+            for phase in self.phase_positions.keys()
+        })
+        self.gmr.update({
+            phase: self.GMR_cn(phase)
+            for phase in model.diameter_over_neutral.keys()
+        })
+        self.r.update({
+            phase: resistance / model.neutral_strand_count[phase]
+            for phase, resistance in model.neutral_strand_resistance.items()
+        })
+        return
+
+    def compute_d(self, i, j):
```

Comment:
# Comparison to Gridlab-d

To try and find the issues with the impedance calculations, I audited the gridlab-d source. I think this method differs from the [Gridlab-d](https://github.com/gridlab-d/gridlab-d/blob/master/powerflow/underground_line.cpp#L403-L450) distance calculations in the following cases.

## Background

Gridlab-d appears to use the following indexes for different phases. The snippets below use these phase indexes

| Index | Description |
| ----- | ----------- |
| 1 | Phase A |
| 2 | Phase B |
| 3 | Phase C |
| 4 | Concentric Neutral A |
| 5 | Concentric Neutral B |
| 6 | Concentric Neutral C |
| 7 | Neutral Conductor | 

## Conductor to Neutral Cable

### Python Implementation
In this case, the inputs are `I = {'A'}, J = {'N'}`

This leads to the following:

```python
I ^ J = {'A', 'N'}
I & J = {}

one_neutral_same_phase == False
different_phase == True
one_neutral == True
```
Thus the code passed the check on line 233 and returns `(distance_ij**2 + r**2) ** 0.5`.

### Gridlab-D Implementation
However, gridlab-d does the following in the case of A -> N:

```cpp
#define DIST(ph1, ph2) (has_phase(PHASE_##ph1) && has_phase(PHASE_##ph2) && config->line_spacing ? OBJECTDATA(config->line_spacing, line_spacing)->distance_##ph1##to##ph2 : 0.0)
D(1, 7) = DIST(A, N);
```

### Solution
I think the correct solution in this case is to return `distance_ij`

## Conductor to Own Concentric Neutral

## Python Implementation
In this case, the inputs are `I = {'A'}, J = {'A', 'N'}`

This leads to the following:

```python
I ^ J = {'N'}
I & J = {'A}

one_neutral_same_phase == True
different_phase == False
one_neutral == True
```

Thus the check on line 227 is true, and so `r` is returned. This value was calculated earlier on lines 200 and 201 as

```python
(diameter_over_neutral - model.neutral_strand_diameter[phase]) / 2
```

### Gridlab-d Distance
Gridlab-d does the following:
```cpp
dia_od1 = UG_GET(A, outer_diameter);
DIA(4) = UG_GET(A, neutral_diameter);
rad_14 = (dia_od1 - DIA(4)) / 24.0;
D(1, 4) = rad_14;
```
### Solution

Divide by 24 on line 201 instead of 2. Additionally, check that the gridlab-d outer diameter is the diameter over the neutral.

## Conductor to Different Phase Concentric Neutral

### Python Implementation

In this case, the inputs are `I = {'A'}, J = {'B', 'N'}`

This leads to the following:

```python
I ^ J = {'A', 'B', 'N'}
I & J = {}

one_neutral_same_phase == False
different_phase == True
one_neutral == True
```

Thus the check on line 233 is True, and so the python code returns `(distance_ij**2 + r**2) ** 0.5`

### Gridlab-d Implementation

```cpp
D(1, 5) = D(1, 2);
```

In otherwords, the distance from A -> BN is the same as the distance from A -> B.

## Concentric Neutral to Another Concentric Neutral

### Python Implementation

In this case, the inputs are `I = {'A', 'N'}, J = {'B', 'N'}`

This leads to the following:

```python
I ^ J = {'A', 'B'}
I & J = {'N'}

one_neutral_same_phase == False
different_phase == False
one_neutral == False
```

Thus the check on line 227 and the check on line 233 are False and so `distance_ij` is returned

### Gridlab-d Implementation
```cpp
D(4, 5) = D(1, 2);
```
In other words, the distance from AN -> BN is the same as the distance from A -> B. The code appears to be correct here

Reply:
Thanks for the thorough audit @etimberg ! A couple points here:

- For case `Conductor to Own Concentric Neutral`, I think gridlab-d divides by `24` to include unit conversion from `inch` to `ft`. We handle everything using metric units, so dividing by `2` to get radius from diameter.

- For case `Conductor to Different Phase Concentric Neutral`, our implementation follows the assumption and example given in the Kersting's book, approximating the concentric neutrals as one single wire directly above their respective phase conductors. So there comes the Pythagorean theorem.

Reply:
That makes sense @veronicaguo re 2 instead of 24. For the `Conductor to Different Phase Concentric Neutral` case, I'm not sure that the distance calculation is correct. It assumes all the cables are in one horizontal plane because that's the only way the triangle becomes right angle. If the cables were arranged in a geometry similar to [this image](https://cdn.comsol.com/wordpress/2017/02/cross-section-of-underground-electrical-cable.png), the distance from the top conductor to one of the bottom concentric neutrals is formed by a non right triangle and so the extra term from the cosine law would need to be added.

Maybe a way to simplify the code here is to calculate the position (x,y) of the concentric neutral in a prior step. Then, all you'd do here is lookup `distance_ij`. It might be worthwhile testing this function independently too for the following cases:
* Horizontal geometry
* Vertical geometry
* Bundled like the image above